pkg/grpc/codec.go

```diff
@@ -0,0 +1,62 @@
+package grpc
+
+import (
+	"google.golang.org/grpc/encoding"
+	"google.golang.org/grpc/encoding/proto"
+	"google.golang.org/grpc/mem"
+)
+
+type vtprotoMessage interface {
+	MarshalToSizedBufferVT(data []byte) (int, error)
+	UnmarshalVT([]byte) error
+	SizeVT() int
+}
+
+type codec struct {
+	// similar to customMarshaler we fall back to original implementation when message is not supported
+	encoding.CodecV2
+}
+
+func (c *codec) Name() string { return c.CodecV2.Name() }
+
+var defaultBufferPool = mem.DefaultBufferPool()
+
+func (c *codec) Marshal(v any) (mem.BufferSlice, error) {
+	m, ok := v.(vtprotoMessage)
+	if !ok {
+		return c.CodecV2.Marshal(v)
+	}
+	size := m.SizeVT()
+	if mem.IsBelowBufferPoolingThreshold(size) {
+		buf := make([]byte, size)
+		_, err := m.MarshalToSizedBufferVT(buf)
+		if err != nil {
+			return nil, err
```

Comment:
On error, try and fallback to the default codec marshaling (also applies to the pooled variant).